### PR TITLE
os/bluestore: try to split blobs instead of spanning them

### PR DESCRIPTION
**Pull request by liewegas**
If a blob crosses a shard boundary, try to split it into pieces
instead of making it spanning.  This is only possible under
certain conditions (uncompressed, no in-flight writes, no unused
space, shard boundary aligned with chunk boundary).

Generally speaking this will reduce the number of blobs that end
up spanning and reduce the onode key size.

Signed-off-by: Sage Weil <sage@redhat.com>

### CODE REVIEW

File: src/os/bluestore/bluestore_types.h [hunk 1]
```diff
@@ -579,6 +579,23 @@ struct bluestore_blob_t {
   int verify_csum(uint64_t b_off, const bufferlist& bl, int* b_bad_off,
 		  uint64_t *bad_csum) const;
 
+  bool can_prune_tail() const {
+    return
+      !extents.empty() &&
```
Comment (ifed01) [inline src/os/bluestore/bluestore_types.h]: can be removed since there is extents.size > 1

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -2026,6 +2158,51 @@ BlueStore::Extent *BlueStore::ExtentMap::set_lextent(
...
+      continue;
+    }
+    if (ep->end() <= pos) {
+      continue;
+    }
+    vector<bluestore_pextent_t> released;
+    if (ep->logical_offset < pos) {
+      // split extent
+      size_t left = pos - ep->logical_offset;
+      Extent *ne = new Extent(pos, 0, ep->length - left, rb);
```
Comment (somnathr) [inline src/os/bluestore/BlueStore.cc]: @liewegas this is giving compilation error in latest master after I cherry-picked. Extent constructor is taking BlobRef& b now.

**Comment by liewegas** [comment]
rebsaed and fixed!

**Comment by somnathr** [comment]
@liewegas Getting the following onode size (in avg) after almost 7 hours of writing.

Put( Prefix = O key = 0x7f80000000000000019390e2c2217262'd_data.10156b8b4567.00000000000275f9!='0xfffffffffffffffeffffffffffffffff001c1000'x' Value size = 626)
Put( Prefix = O key = 0x7f80000000000000019390e2c2217262'd_data.10156b8b4567.00000000000275f9!='0xfffffffffffffffeffffffffffffffff Value size = 663)

This is encouraging, combining all those recent sharding/blob changes I am getting a total of ~65% performance improvement.

855 bytes of pg_info is significant now , I am hoping for another 35-40% improvement when we cut it to 100 bytes.

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -1583,10 +1683,42 @@ void BlueStore::ExtentMap::reshard(Onode *o)
...
+		dout(20) << __func__ << "    splitting blob, bstart 0x"
+			 << std::hex << bstart
+			 << " blob_offset 0x" << blob_offset
+			 << std::dec << " " << *b << dendl;
+		b = split_blob(b, blob_offset, sh.offset);
+		// switch b to the new right-hand side, in case it
+		// *also* has to get split.
+		bstart += blob_offset;
+	      } else {
+		must_span = true;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: break the loop here?

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -2026,6 +2158,51 @@ BlueStore::Extent *BlueStore::ExtentMap::set_lextent(
...
+  BlobRef rb = onode->c->new_blob();
+  lb->split(blob_offset, rb.get());
+
+  for (auto ep = seek_lextent(pos);
+       ep != extent_map.end() && ep->logical_offset < end_pos;
+       ++ep) {
+    if (ep->blob != lb) {
+      continue;
+    }
+    if (ep->end() <= pos) {
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: IMHO that's an overkill since seek_lextent takes care of it.

**Comment by liewegas** [comment]
updated, thanks!

File: src/os/bluestore/BlueStore.cc [hunk 5]
```diff
@@ -1382,6 +1426,62 @@ bool BlueStore::Blob::put_ref(
...
+  unsigned i = 0;
+  size_t left = blob_offset;
+  for (auto p = lb.extents.begin(); p != lb.extents.end(); ++p, ++i) {
+    if (p->length <= left) {
+      left -= p->length;
+      continue;
+    }
+    if (left) {
+      if (p->is_valid()) {
+	rb.extents.emplace_back(bluestore_pextent_t(p->offset + left,
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: I'm worrying if we can have pextent with arbitrary (not aligned with min_alloc_size) boundaries. Such pextent might be eventually released and I doubt allocator handles this properly. Or I missed something? Anyway suggest to have corresponding test case in ceph_test_objectstore

**Comment by liewegas** [comment]
good call, added a patch to prevent that.

File: src/os/bluestore/BlueStore.cc [hunk 6]
```diff
@@ -1701,7 +1701,8 @@ void BlueStore::ExtentMap::reshard(Onode *o)
 	  for (const auto& sh : shards) {
 	    if (bstart < sh.offset && bend > sh.offset) {
 	      uint32_t blob_offset = sh.offset - bstart;
-	      if (b->get_blob().can_split_at(blob_offset)) {
+	      if (b->get_blob().can_split_at(blob_offset) &&
+		  blob_offset % min_alloc_size == 0) {
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: And another concern - are there any means that ensure blob_offset to be aligned with min_alloc_size often enough. If no and blob_offset is pretty arbitrary then split blob functionality makes no sense...

**Comment by liewegas** [comment]
In practice, with block devices, the extent boundaries will fall on the 
alignment boundaries.  Probably less true with CephFS objects.  We 
probably need to make this capable of splitting extents too.  I think we 
can do that later, though.. the main piece of infrastructure I'm most 
interested here is the split_blob part.  reshard() can be improved as we 
go.

**Comment by ifed01** [comment]
LGTM

**Pull request merged**